aries_cloudagent/messaging/valid.py

```diff
@@ -17,6 +19,52 @@
 B58 = alphabet if isinstance(alphabet, str) else alphabet.decode("ascii")
 
 
+class StrOrDictField(Field):
+    """URI or Dict field for Marshmallow."""
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        return value
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        if isinstance(value, (str, dict)):
+            return value
+        else:
+            raise ValidationError("Field should be str or dict")
+
+
+class DictOrDictListField(Field):
+    """Dict or Dict List field for Marshmallow."""
+
+    def _serialize(self, value, attr, obj, **kwargs):
+        return value
+
+    def _deserialize(self, value, attr, data, **kwargs):
+        # dict
+        if isinstance(value, dict):
+            return value
+        # list of dicts
+        elif isinstance(value, list) and all(isinstance(item, dict) for item in value):
+            return value
+        else:
+            raise ValidationError("Field should be dict or list of dicts")
+
+
+class UriOrDictField(StrOrDictField):
+    """URI or Dict field for Marshmallow."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize new UriOrDictField instance."""
+        super().__init__(*args, **kwargs)
+
+        # Insert validation into self.validators so that multiple errors can be stored.
+        self.validators.insert(0, self._uri_validator)
+
+    def _uri_validator(self, value):
+        # Check if URI when
+        if isinstance(value, str):
+            return Uri()(value)
+
+
```

Comment:
I'm going to steal this code and put it into its own (new) file in messaging/field.py. It will show up in https://github.com/hyperledger/aries-cloudagent-python/pull/1080.